runtime/runtime-params-estimator/src/replay/fold_db_ops.rs

```diff
@@ -0,0 +1,268 @@
+    cache_stats: Option<CacheStats>,
+}
+
+impl FoldDbOps {
+    pub(super) fn new() -> Self {
+        Self {
+            fold_anchors: HashMap::new(),
+            print_top_level: true,
+            account_filter: None,
+            track_caches: false,
+            states: vec![State::default()],
+            block_hash: None,
+            skip_indent: None,
+        }
+    }
+
+    /// Pre-set that folds on chunks.
+    pub(super) fn chunks(self) -> Self {
+        self.fold("apply_transactions", &["receiver", "shard_id"])
+    }
+
+    /// Pre-set that folds on receipts.
+    pub(super) fn receipts(self) -> Self {
+        self.fold("process_receipt", &["receiver", "receipt_id"])
+            .fold("process_transaction", &["receipt_id"])
+            .print_top_level(false)
+    }
+
+    pub(super) fn fold(mut self, anchor: impl Into<String>, printed_fields: &[&str]) -> Self {
+        let fields = printed_fields.into_iter().map(|s| (*s).into()).collect();
+        self.fold_anchors.insert(anchor.into(), fields);
+        self
+    }
+
+    pub(super) fn print_top_level(mut self, yes: bool) -> Self {
+        self.print_top_level = yes;
+        self
+    }
+
+    pub(super) fn with_cache_stats(mut self) -> Self {
+        self.track_caches = true;
+        self.state().cache_stats = Some(CacheStats::default());
+        self
+    }
+
+    pub(super) fn account_filter(mut self, account: Option<String>) -> Self {
+        self.account_filter = account;
+        self
+    }
+
+    fn state(&mut self) -> &mut State {
+        self.states.last_mut().expect(EMPTY_STATE_ERR)
+    }
+
+    fn push_state(&mut self, indent: usize) {
+        let cache_stats = if self.track_caches { Some(CacheStats::default()) } else { None };
+        let new_state = State { indent, ops_cols: Default::default(), cache_stats };
+        self.states.push(new_state);
+    }
+
+    fn pop_state(&mut self) -> State {
+        let state = self.states.pop().expect(EMPTY_STATE_ERR);
+        if self.states.is_empty() {
+            self.push_state(0);
+        }
+        state
+    }
+
+    fn skip(&mut self, indent: usize) -> bool {
+        if let Some(skip_indent) = self.skip_indent {
+            if skip_indent >= indent {
+                self.skip_indent = None;
+            } else {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Check if indentation has gone back enough to pop current state.
+    fn check_indent(&mut self, out: &mut dyn Write, indent: usize) -> anyhow::Result<()> {
+        if self.states.len() > 1 && self.state().indent >= indent {
+            self.pop_state().print(out)?;
+        }
+        Ok(())
+    }
+}
+
+impl Visitor for FoldDbOps {
+    fn eval_db_op(
+        &mut self,
+        out: &mut dyn Write,
+        indent: usize,
+        op: &str,
+        size: Option<u64>,
+        key: &[u8],
+        col: &str,
+    ) -> anyhow::Result<()> {
+        self.check_indent(out, indent)?;
```

Comment:
I took another look and now I don't understand why here `check_indent` comes before `skip`, it is the same for `eval_storage_op`, but for `eval_label` the order is different.

Could you explain this logic? Maybe `fn skip` is a good place to comment this behaviour. Similarly, I think we need some test which fails for the case when order is incorrect.